src/XrdVoms/XrdVomsMapfile.cc

```diff
@@ -0,0 +1,359 @@
+{
+    if (entry.m_path.empty()) {return false;}
+
+    // A more specific mapfile entry cannot match a generic FQAN
+    if (fqan.size() < entry.m_path.size()) {return false;}
+
+    XrdOucString fqan_element;
+    for (size_t idx=0; idx<entry.m_path.size(); idx++) {
+        fqan_element.assign(fqan[idx].c_str(), 0);
+        const auto &path_element = entry.m_path[idx];
+        if (!fqan_element.matches(path_element.c_str())) {return false;}
+    }
+    if (fqan.size() == entry.m_path.size()) {return true;}
+    if (entry.m_path.back() == "*") {return true;}
+    return false;
+}
+
+
+std::vector<std::string>
+XrdVomsMapfile::MakePath(const XrdOucString &group)
+{
+    int from = 0;
+    XrdOucString entry;
+    std::vector<std::string> path;
+    path.reserve(4);
+    while ((from = group.tokenize(entry, from, '/')) != -1) {
+        if (entry.empty()) continue;
+        path.emplace_back(entry.c_str());
+    }
+    return path;
+}
+
+
+int
+XrdVomsMapfile::GetSecData(XrdLink * lnk, XrdSecEntity &entity, SSL *ssl)
+{
+    if (!m_xrdvoms) return -1;
+
+    auto retval = m_xrdvoms->GetSecData(lnk, entity, ssl);
+    if (retval) return retval;
+
+    return Apply(entity);
+}
+
+
+int
+XrdVomsMapfile::Apply(XrdSecEntity &entity)
+{
+    Reconfigure();
+
+    int from_vorg = 0, from_role = 0, from_grps = 0;
+    XrdOucString vorg = entity.vorg, entry_vorg;
+    XrdOucString role = entity.role, entry_role;
+    XrdOucString grps = entity.grps, entry_grps;
+    if (m_edest) m_edest->Log(LogMask::Debug, "VOMSMapfile", "Applying VOMS mapfile to incoming credential");
+    while (((from_vorg = vorg.tokenize(entry_vorg, from_vorg, ' ')) != -1) &&
+           ((from_role = role.tokenize(entry_role, from_role, ' ')) != -1) &&
+           ((from_grps = grps.tokenize(entry_grps, from_grps, ' ')) != -1))
+    {
+        auto fqan = MakePath(entry_grps);
+        if (fqan.empty()) {continue;}
+
+        // By convention, the root group should be the same as the VO name; however,
+        // the VOMS mapfile makes this assumption.  To be secure, enforce it.
+        if (strcmp(fqan[0].c_str(), entry_vorg.c_str())) {continue;}
+
+        fqan.emplace_back(std::string("Role=") + entry_role.c_str());
+        fqan.emplace_back("Capability=NULL");
+        std::string username;
+        if (!(username = Map(fqan)).empty()) {
+            if (entity.name) {free(entity.name);}
+            entity.name = strdup(username.c_str());
+            break;
+        }
+    }
+
+    return 0;
+}
+
+
+XrdVomsMapfile *
+XrdVomsMapfile::Get()
+{
+    return mapper.get();
+}
+
+
+XrdVomsMapfile *
+XrdVomsMapfile::Configure(XrdSysError *erp, XrdHttpSecXtractor *xtractor)
+{
+    if (tried_configure) {
+        auto result = mapper.get();
+        if (result) {
+            result->SetExtractor(xtractor);
+            result->SetErrorStream(erp);
+        }
+        return result;
+    }
+
+    tried_configure = true;
```

Comment:
Why do we need this at all. The assumption, code-wise, is that this will be called during initialization and hence will be called only once. So, it's very confusing why this code exists here at all. Is there something we don't know about?

Reply:
It's necessary in the current setup: the two protocols are configured independently.  Hence, `XrdVoms` is invoked twice whereas i want the mapper to be a singleton.

Reply:
OK, could you please add a comment for why the code is actually there?

Reply:
Comment is added in the latest set of commits.  Resolving.